components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/handler/UserSelfSignUpHandler.java

```diff
@@ -0,0 +1,232 @@
+                                                  secretKey,
+                                                arbitraryProperties);
+                    } else {
+                        eventContext.addParameter(IdentityStoreConstants.UNIQUE_USED_ID, user.getUniqueUserId());
+                        eventContext.addParameter(CONFIRMATION_CODE, secretKey);
+                    }
+                } else {
+
+                    if (log.isDebugEnabled()) {
+                        log.debug("Changing user lifecycle state to {} for user: {}", UserState
+                                .UNLOCKED__UNVERIFIED.toString(), user.getUniqueUserId());
+                    }
+                    LifecycleOperationManager.executeLifecycleEvent(UserState.UNLOCKED__UNVERIFIED.toString(),
+                                                                    domainUID, domainUID, null);
+                    identityStore.setUserState(user.getUniqueUserId(), UserState.UNLOCKED__UNVERIFIED.toString());
+                }
+            } catch (LifecycleException e) {
+                throw Utils.handleServerException(ErrorCodes.FAILED_LIFECYCLE_EVENT, null, e);
+            } catch (UserNotFoundException e) {
+                throw Utils.handleServerException(ErrorCodes.USER_NOT_FOUND, null, e);
+            } catch (IdentityStoreException e) {
+                throw Utils.handleServerException(ErrorCodes.FAILED_USER_STATE_UPDATE, null, e);
+            }
+        }
+
+        if (IdentityStoreInterceptorConstants.POST_AUTHENTICATE.equals(event.getEventName())) {
+
+            Map<String, Object> eventProperties = event.getEventProperties();
+
+            AuthenticationContext authenticationContext = (AuthenticationContext) eventProperties.get
+                    (IdentityStoreConstants.AUTHENTICATION_CONTEXT);
+
+            if (authenticationContext.isAuthenticated()) {
+
+                User user = authenticationContext.getUser();
+                if (UserState.LOCKED_SELF_SIGN_UP.toString().equals(user.getState())) {
+                    throw Utils.handleClientException(ACCOUNT_UNVERIFIED, null);
+                }
+            }
+        }
+    }
+
+    @Override
+    public int getPriority(MessageContext messageContext) {
+        // TODO: To be removed after introducing handler config properties
+        return -10;
+    }
+
+    /**
+     * This is used to extract domainId from unique user id.
+     *
+     * @param uniqueEntityId Unique user id.
+     * @return Domain Id.
+     * @throws EventException If the Unique ID is invalid.
+     */
+    private String getDecodedUserEntityId(String uniqueEntityId) throws EventException {
+
+        String[] decodedUniqueEntityIdParts = uniqueEntityId.split("\\.", 2);
+        if (decodedUniqueEntityIdParts.length != 2 || StringUtils.isEmpty(decodedUniqueEntityIdParts[0]) ||
+            StringUtils.isEmpty(decodedUniqueEntityIdParts[1])) {
+
+            throw new EventException("Invalid unique user id.");
```

Comment:
Better to put uniqueEntityId on the exception if it is not sensitive. Otherwise we need to say something like "unique user id format is not recognized. It should be in the format xxx\yyy, but was AB***YZ.